flow/BlobCipher.cpp

```diff
@@ -0,0 +1,392 @@
+                             const BlobCipherBaseKeyId& baseCiphId,
+                             const uint8_t* baseCiph,
+                             int baseCiphLen) {
+	BlobCipherRandomSalt salt;
+	if (g_network->isSimulated()) {
+		salt = deterministicRandom()->randomUInt64();
+	} else {
+		salt = nondeterministicRandom()->randomUInt64();
+	}
+	initKey(domainId, baseCiph, baseCiphLen, baseCiphId, salt);
+	/*TraceEvent("BlobCipherKey")
+	    .detail("DomainId", domainId)
+	    .detail("BaseCipherId", baseCipherId)
+	    .detail("BaseCipherLen", baseCipherLen)
+	    .detail("RandomSalt", randomSalt)
+	    .detail("CreationTime", creationTime);*/
+}
+
+void BlobCipherKey::initKey(const BlobCipherDomainId& domainId,
+                            const uint8_t* baseCiph,
+                            int baseCiphLen,
+                            const BlobCipherBaseKeyId& baseCiphId,
+                            const BlobCipherRandomSalt& salt) {
+	// Set the base encryption key properties
+	baseCipher = std::make_unique<uint8_t[]>(AES_256_KEY_LENGTH);
+	memset(baseCipher.get(), 0, AES_256_KEY_LENGTH);
+	memcpy(baseCipher.get(), baseCiph, std::min<int>(baseCiphLen, AES_256_KEY_LENGTH));
+	baseCipherLen = baseCiphLen;
+	baseCipherId = baseCiphId;
+	// Set the encryption domain for the base encryption key
+	encryptDomainId = domainId;
+	randomSalt = salt;
+	// derive the encryption key
+	cipher = std::make_unique<uint8_t[]>(AES_256_KEY_LENGTH);
+	memset(cipher.get(), 0, AES_256_KEY_LENGTH);
+	applyHmacSha256Derivation();
+	// update the key creation time
+	creationTime = now();
+}
+
+void BlobCipherKey::applyHmacSha256Derivation() {
+	Arena arena;
+	uint8_t buf[baseCipherLen + sizeof(BlobCipherRandomSalt)];
+	memcpy(&buf[0], baseCipher.get(), baseCipherLen);
+	memcpy(&buf[0] + baseCipherLen, &randomSalt, sizeof(BlobCipherRandomSalt));
+	HmacSha256DigestGen hmacGen(baseCipher.get(), baseCipherLen);
+	StringRef digest = hmacGen.digest(&buf[0], baseCipherLen + sizeof(BlobCipherRandomSalt), arena);
+	std::copy(digest.begin(), digest.end(), cipher.get());
+	if (digest.size() < AES_256_KEY_LENGTH) {
+		memcpy(cipher.get() + digest.size(), buf, AES_256_KEY_LENGTH - digest.size());
+	}
+}
+
+void BlobCipherKey::reset() {
+	memset(baseCipher.get(), 0, baseCipherLen);
+	memset(cipher.get(), 0, AES_256_KEY_LENGTH);
+}
+
+// BlobCipherKeyCache class methods
+
+void BlobCipherKeyCache::insertCipherKey(const BlobCipherDomainId& domainId,
+                                         const BlobCipherBaseKeyId& baseCipherId,
+                                         const uint8_t* baseCipher,
+                                         int baseCipherLen) {
+	Reference<BlobCipherKey> cipherKey =
+	    makeReference<BlobCipherKey>(domainId, baseCipherId, baseCipher, baseCipherLen);
+
+	auto domainItr = domainCacheMap.find(domainId);
+	if (domainItr == domainCacheMap.end()) {
+		domainCacheMap[domainId].emplace(baseCipherId, makeReference<BlobCipherKeyItem>(cipherKey));
```

Comment:
Is there a reason that the key cache and domain cache have 2 different references to the cipher key? The only one that is apparent to me is that the domain copy and the key copy would have their own BlobCipherKeySaltCache, but that doesn't necessarily seem desirable?
Additionally, this means we'd call updateLatest on only the key map's BlobCipherKeyItem? Which seems incorrect

Reply:
Am not sure if I understood the question here. Lets take this offline, may be easier to reason through it.